wiring/inc/spark_wiring_diagnostics.h

```diff
@@ -0,0 +1,367 @@
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include "spark_wiring_thread.h"
+#include "spark_wiring_interrupts.h"
+#include "spark_wiring_error.h"
+#include "spark_wiring_global.h"
+
+#include "diagnostics.h"
+
+#include <atomic>
+
+namespace particle {
+
+// Base abstract class for a diagnostic data source
+class AbstractDiagnosticData {
+public:
+    static int get(uint16_t id, void* data, size_t& size);
+    static int get(const diag_source* src, void* data, size_t& size);
+
+    // This class is non-copyable
+    AbstractDiagnosticData(const AbstractDiagnosticData&) = delete;
+    AbstractDiagnosticData& operator=(const AbstractDiagnosticData&) = delete;
+
+protected:
+    AbstractDiagnosticData(uint16_t id, diag_type type);
+    AbstractDiagnosticData(uint16_t id, const char* name, diag_type type);
+
+    virtual int get(void* data, size_t& size) = 0;
+
+private:
+    diag_source d_;
+
+    static int callback(const diag_source* src, int cmd, void* data);
+};
+
+// Base abstract class for a data source containing an integer value
+class AbstractIntegerDiagnosticData: public AbstractDiagnosticData {
+public:
+    typedef int32_t IntType; // Underlying integer type
+
+    static int get(uint16_t id, IntType& val);
+    static int get(const diag_source* src, IntType& val);
+
+protected:
+    explicit AbstractIntegerDiagnosticData(uint16_t id, const char* name = nullptr);
+
+    virtual int get(IntType& val) = 0;
+
+private:
+    virtual int get(void* data, size_t& size) override; // AbstractDiagnosticData
+};
+
+template<typename LockingPolicyT>
+class IntegerDiagnosticData: public AbstractIntegerDiagnosticData, private LockingPolicyT {
+public:
+    explicit IntegerDiagnosticData(uint16_t id, IntType val = 0);
+    IntegerDiagnosticData(uint16_t id, const char* name, IntType val = 0);
+
+    IntType operator++();
+    IntType operator++(int);
+    IntType operator--();
+    IntType operator--(int);
+    IntType operator+=(IntType val);
+    IntType operator-=(IntType val);
+
+    IntegerDiagnosticData<LockingPolicyT>& operator=(IntType val);
+    operator IntType() const;
+
+private:
+    IntType val_;
+
+    virtual int get(IntType& val) override; // AbstractIntegerDiagnosticData
+};
+
+template<>
+class IntegerDiagnosticData<AtomicLockingPolicy>: public AbstractIntegerDiagnosticData {
+public:
+    explicit IntegerDiagnosticData(uint16_t id, IntType val = 0);
+    IntegerDiagnosticData(uint16_t id, const char* name, IntType val = 0);
+
+    IntType operator++();
+    IntType operator++(int);
+    IntType operator--();
+    IntType operator--(int);
+    IntType operator+=(IntType val);
+    IntType operator-=(IntType val);
+
+    IntegerDiagnosticData<AtomicLockingPolicy>& operator=(IntType val);
+    operator IntType() const;
+
+private:
+    std::atomic<IntType> val_;
```

Comment:
What's the overhead of using std:atomic here with SimpleIntegerDiagnosticData?  The majority of cases the diagnostic data doesn't need to be thread safe, only a few cases need it, so the non-atomic version should be the more efficient case ideally.

Reply:
`SimpleIntegerDiagnosticData` uses a dummy [`NoLockingPolicy`](https://github.com/spark/firmware/blob/feature/diagnostics/wiring/inc/spark_wiring_thread.h#L236), so there shouldn't any overhead. The class in question is a specialization of `IntegerDiagnosticData` for `AtomicLockingPolicy`, i.e. it's up to a developer to decide if a particular diagnostic data needs to be updated atomically.